### ISSUE TITLE
Back button behavior in search mode

### ISSUE THREAD

**Issue by ltGuillaume** [TBog/TBLauncher#114]
1. ~~I think that clearing the search term on the first back press isn't intuitive: people who would like to see the list on the results entire screen would most likely press the back key, as this normally hides the keyboard in almost every app, _without changing the input_.~~
2. Also, I'm afraid it would be very confusing for new users that the second time you press the back button, the launcher won't be returning to the previous desktop mode (widget/empty). What's your take on this?

**Comment by TBog** [comment]
> 2. Also, I'm afraid it would be very confusing for new users that the second time you press the back button, the launcher won't be returning to the previous desktop mode (widget/empty). What's your take on this?

I understand what you mean but the 🔙 button is meant to be going back when you have a progression. Going from Search to Widgets to Empty is not a progression imho. 

I can make the 🔙 button do
1. the opposite of the Tap empty screen when Toggle is set as a gesture 
2. have a setting for what the back button should do in each desktop

What do you prefer?

**Comment by ltGuillaume** [comment]
> I understand what you mean but the 🔙 button is meant to be going back when you have a progression. Going from Search to Widgets to Empty is not a progression imho.

I understand, but that's not what I meant. I meant only going back to the previous mode. So, upon entering the search mode, you could just save "prevMode" or so, and the back key would reinstate that mode.

Haven't looked at the code, but _if_ the modes are separate activities, just add `someIntent.putExtra("prevMode", 1);` for widgets, for example, and pick it up in the search mode.

> I can make the 🔙 button do
> 
>     1. the opposite of the Tap empty screen when Toggle is set as a gesture
> 
>     2. have a setting for what the back button should do in each desktop
> 
> 
> What do you prefer?

2 would also do, but I don't think a separate setting is necessary if my previous suggestion would be implemented.

**Comment by TBog** [comment]
So if I am going from Widgets to Search pressing 🔙 would open Widgets again.
Now I am in Widgets and I press 🔙, what should open?

What about the Empty mode?

**Comment by ltGuillaume** [comment]
> So if I am going from Widgets to Search pressing 🔙 would open Widgets again.
> Now I am in Widgets and I press 🔙, what should open?

Nothing! You're back where you started!

> What about the Empty mode?

If you go from Empty to Search, pressing 🔙 would open Empty again.

**Comment by TBog** [comment]
Aha, you assume we have a starting screen.
Which is that screen and how did you determine it?

**Comment by ltGuillaume** [comment]
No, I'm not assuming that. I guess that some people would like to see the Empty screen when they get back to their launcher, while other - like me - would like to see the Widgets screen.

Also, I think that's besides the point. I'm just advocating the simple and most intuitive workflow of "if you go into Search from the Widgets, then 🔙 should take you to Widgets again; if you go into Search from the Empty, then 🔙 should take you to Empty again". That's all.

**Comment by ltGuillaume** [comment]
Just to be clear, it did work like that before... I know the new "Desktop mode doctrine" doesn't dictate that it should still do so, but it is definitely the most intuitive way.

**Comment by TBog** [comment]
> Just to be clear, it did work like that (2) before...

It was inherited from KISS that pressing back would hide the search bar. That meant showing the widgets in the old design. Now we have 3 modes.

**Comment by ltGuillaume** [comment]
Ok, I get your point that desktop modes should be able to be completely separate. In that case, it might be best to implement your second suggestion (have a setting for what the back button should do in each desktop) to be versatile enough to be able to adhere to the more classic interpretation of a launcher, too.

**Comment by ltGuillaume** [comment]
> Aha, you assume we have a starting screen.

To get back to this once more: when I close TBLauncher v6 and it automatically relaunches (it has become my only launcher), I automatically get into the Empty mode after some seconds. Does that not make it the starting screen? So there is a starting screen... _There is an option **Widget mode after launch**, which I have enabled, but it doesn't work_.

By the way, before opening the first mode

**Comment by ltGuillaume** [comment]
With the new prerelease, the back button is really funny. I go from Widgets to Search, but end up in Empty when pressing back. Makes no sense.

**Comment by TBog** [comment]
You can change in the settings the behavior of the 🔙 button, wouldn't that fix the problem?

**Comment by ltGuillaume** [comment]
I'm so sorry, I expected there to be a new setting directly under Features, but there's a new one under each desktop mode. Totally missed that!

**Comment by TBog** [comment]
Is this fixed in [v6. 0-dev.1](https://github.com/TBog/TBLauncher/releases/tag/v6.0-dev.1) ?

**Comment by ltGuillaume** [comment]
Nope, I checked beforehand.

**Comment by TBog** [comment]
> Nope, I checked beforehand.

The back button should respect your choice from settings. What is not working?

**Comment by TBog** [comment]
> By the way, before opening the first mode there seems to be a hardcoded semitransparent background for the status bar:
 fixed in 759c5319b48581e9fd035a7bade0b087ca74ffbf

**Comment by ltGuillaume** [comment]
The semitransparent background behind the status bar remains, but it does open the empty mode by default now. 


> It was not the empty screen, it was an uninitialized state,

Yes, that's what I figured, as described after the sentence you quoted.

> that's why I told you I had no starting screen to go back to.

Fair enough, I was only thinking about the back button behavior in the search mode in this issue (as in title), compared to previous versions.

The first time TBLauncher starts, it's fine if it opens in Empty mode. I'm gonna test its behavior when Android pushes it out of RAM: it would be expected to return to the previous mode. Otherwise, perhaps it would be a great addition to allow the user to define the start-up mode.

**Comment by TBog** [comment]
> The semitransparent background behind the status bar remains, but it does open the empty mode by default now.

I can't reproduce this in the emulator or my phone. Maybe it's a setting you have?
Need help to solve this.

**Comment by ltGuillaume** [comment]
> 
> 
> > The semitransparent background behind the status bar remains, but it does open the empty mode by default now.
> 
> I can't reproduce this in the emulator or my phone. Maybe it's a setting you have?
> Need help to solve this.
> 
> I'll close this issue as the initial issue is solved. If you still have some problem please open one issue for each.

Well, it's before any of the settings under `User Interface`->`Notification bar` kick in: I put the color to bright red and the opacity to both 100% and 0%, toggled both gradient and dark icons, but in all cases it first showed a dark semitransparent background for the status bar with white icons before those settings were applied. Perhaps there's loading state that uses Android defaults or something?

**Comment by ltGuillaume** [comment]
I just found one minor issue with this:

1. Open search
2. Press back, the keyboard closes
3. Press back again, the action defined for back button is triggered

This is great. Now:

1. Open search
2. Scroll the results, so that the keyboard automatically disappears
3. Press back, *nothing happens*
4. Press back again, the action defined for back button is triggered

So in the second situation, step 3 should be avoided.

**Comment by TBog** [comment]
> So in the second situation, step 3 should be avoided.

It doesn't happen on my device, I'll try on the emulator.

**Comment by ltGuillaume** [comment]
> 
> 
>     1. ~I think that clearing the search term on the first back press isn't intuitive: people who would like to see the list on the results entire screen would most likely press the back key, as this normally hides the keyboard in almost every app, _without changing the input_.~

I already crossed this out before it got implemented, but I was too afraid to ask to revert it 😟 But the fact that the back button first hides the keyboard (and thus doesn't go back to whatever you set it to the first time)...

1. now really really annoys the crap out of me 😅
2. may not align with the "intuitiveness" argument I made before, now there's a per-desktop setting that specifies the behavior of the back button (because now _it actually doesn't_ do right away what you explicitly specified for it to do).
